File: Python/16. 3Sum Closest.py
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest = (10000,0)
        for i,a in enumerate(nums):
            j,k = i+1,len(nums)-1
            while j<k:
                tsum = a + nums[j] + nums[k]
                sumdistance = tsum-target
                if abs(sumdistance)<closest[0]:
                    closest = (abs(sumdistance),tsum)
                    if sumdistance<0:
                        j+=1
                    elif sumdistance>0:
                        k-=1
                    else:
                        return tsum
                else:
                    if sumdistance<0:
                        j+=1
                    elif sumdistance>0:
                        k-=1

        return closest[1]
    
if __name__ == "__main__":
    nums = [-1,2,1,-4]
    target = 1
    result = Solution().threeSumClosest(nums=nums,target=target)
    print(result)

File: CPP/16. 3Sum Closest.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

class Solution {
public:
	int threeSumClosest(vector<int> nums, int target) {
		sort(nums.begin(), nums.end());
		pair<int, int> closest = { 10000,0 };

		for (size_t i = 0; i < nums.size(); i++) {
			size_t j = i + 1, k = nums.size() - 1;
			while (j < k) {
				int tsum = nums[i] + nums[j] + nums[k];
				int sumdistance = tsum - target;
				if (abs(sumdistance) < closest.first) {
					closest = { abs(sumdistance),tsum };
				}
				if (sumdistance < 0) {
					j++;
				}
				else if(sumdistance > 0) {
					k--;
				}
				else {
					return tsum;
				}
			}
		}
			return closest.second;
	}
};

int main() {
	Solution sol;
	vector<int> nums = { -1, 2, 1, -4 };
	int target = 1;
	cout << sol.threeSumClosest(nums, target) << endl;
	return 0;
}